CA1859 recommends change that fails due to inconsistent accessibility

### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859): Change return type of method M from X to Y for improved performance.

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: 8.0.100-preview.4.23260.5

_and also_

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: [8.0.0-preview1.23165.1](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers/8.0.0-preview1.23165.1) (Latest preview)

<!--
NOTE: `Microsoft.CodeAnalysis.FxCopAnalyzers` package has been deprecated in favor of 'Microsoft.CodeAnalysis.NetAnalyzers', that ships with the .NET SDK.
      Please refer to https://learn.microsoft.com/visualstudio/code-quality/migrate-from-fxcop-analyzers-to-net-analyzers to migrate to .NET analyzers.
-->

### Describe the bug

CA1859 is triggered on code that returns a nested type. The diagnostic message is `Change return type of method 'Method' from 'System.IDisposable' to 'NestedReturnType.Program.NestedClass1.NestedClass2' for improved performance`. If the change is made, the updated code fails to compile with the error `error CS0050: Inconsistent accessibility: return type 'Program.NestedClass1.NestedClass2' is less accessible than method 'Program.NestedClass1.Method()'`.

### Steps To Reproduce

Compile the following code.

<details>
  <summary>CSPROJ file</summary>

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="8.0.0-preview1.23165.1">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```
</details>

```csharp
namespace NestedReturnType;

public class Program
{
	public static void Main(string[] args)
	{
	}

	private static class NestedClass1
	{
		public static IDisposable Method()
		{
			return new NestedClass2();
		}

		private sealed class NestedClass2 : IDisposable
		{
			public void Dispose()
			{
			}
		}
	}
}
```

### Expected behavior

No CA1859 warning.

### Actual behavior

CA1859 warning that recommends an invalid change.

## Additional context

Possibly a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/6479, although .NET SDK 8 Preview 4 and the latest NuGet package have both been released since then, so I expected them to contain the fix.

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs
@@ -154,7 +154,7 @@ namespace Microsoft.NetCore.Analyzers.Performance
                         publicOrInternalColl.ExtractNonPrivate(coll);
 
                         // based on what we've collected, spit out relevant diagnostics for private symbols
-                        Report(context.ReportDiagnostic, coll);
+                        Report(context.ReportDiagnostic, coll, context.Compilation);
                         Collector.ReturnInstance(coll, context.CancellationToken);
                     });
                 }, SymbolKind.NamedType);
@@ -162,7 +162,7 @@ namespace Microsoft.NetCore.Analyzers.Performance
                 context.RegisterCompilationEndAction(context =>
                 {
                     // based on what we've collected, spit out relevant diagnostics for public or internal symbols
-                    Report(context.ReportDiagnostic, publicOrInternalColl);
+                    Report(context.ReportDiagnostic, publicOrInternalColl, context.Compilation);
                     Collector.ReturnInstance(publicOrInternalColl, context.CancellationToken);
                 });
             });
@@ -171,7 +171,7 @@ namespace Microsoft.NetCore.Analyzers.Performance
         /// <summary>
         /// Given all the accumulated analysis state, generate the diagnostics.
         /// </summary>
-        private static void Report(Action<Diagnostic> reportDiag, Collector coll)
+        private static void Report(Action<Diagnostic> reportDiag, Collector coll, Compilation compilation)
         {
             // for all eligible fields that are used as the receiver for a virtual call
             foreach (var pair in coll.VirtualDispatchFields)
@@ -274,22 +274,59 @@ namespace Microsoft.NetCore.Analyzers.Performance
                 {
                     foreach (var t in targets)
                     {
-                        foreach (var m in toType.GetMembers())
+                        var check = toType;
+                        while (check != null)
                         {
-                            if (m.IsImplementationOfAnyExplicitInterfaceMember())
+                            foreach (var m in check.GetMembers())
                             {
-                                if (m.IsImplementationOfInterfaceMember(t))
+                                if (m.IsImplementationOfAnyExplicitInterfaceMember())
                                 {
-                                    return;
+                                    if (m.IsImplementationOfInterfaceMember(t))
+                                    {
+                                        return;
+                                    }
                                 }
                             }
+
+                            check = check.BaseType;
                         }
                     }
                 }
 
-                if (toType.TypeKind != TypeKind.Class)
+                // if the toType or any of its base types introduce methods with the same name as any of the target methods,
+                // we shouldn't recommend to upgrade the type. This is because these new overloads can lead to binding to the
+                // wrong methods in the case of an upgrade from a base type to a derived type.
+                if (targets != null && fromType.TypeKind != TypeKind.Interface)
                 {
-                    // we only deal with classes
+                    using var targetNames = PooledHashSet<string>.GetInstance(targets.Select(t => t.Name));
+                    var check = toType;
+                    while (check != null && check != fromType)
+                    {
+                        foreach (var m in check.GetMembers())
+                        {
+                            if (m is IMethodSymbol ms)
+                            {
+                                if (!ms.IsDefinition || ms.IsOverride)
+                                {
+                                    // those are OK, they won't cause trouble
+                                    continue;
+                                }
+
+                                if (targetNames.Contains(ms.Name))
+                                {
+                                    // OK, we found a match, so we're giving up on this potential upgrade
+                                    return;
+                                }
+                            }
+                        }
+
+                        check = check.BaseType;
+                    }
+                }
+
+                if (toType.TypeKind is not TypeKind.Class and not TypeKind.Array)
+                {
+                    // we only deal with classes or arrays
                     return;
                 }
 
@@ -311,7 +348,7 @@ namespace Microsoft.NetCore.Analyzers.Performance
                     return;
                 }
 
-                if (!toType.GetResultantVisibility().IsAtLeastAsVisibleAs(affectedSymbol.GetResultantVisibility()))
+                if (!HasEquivalentOrGreaterVisibilityToSymbol(compilation, toType, affectedSymbol))
                 {
                     // the suggested type must have equal or greater visibility than the affected symbol.
                     return;
@@ -321,6 +358,37 @@ namespace Microsoft.NetCore.Analyzers.Performance
                 var toTypeName = GetTypeName(toType);
                 var diagnostic = affectedSymbol.CreateDiagnostic(desc, affectedSymbol.Name, fromTypeName, toTypeName);
                 reportDiag(diagnostic);
+            }
+
+            // ensures that the type can be referenced from any code that can also reference the symbol
+            static bool HasEquivalentOrGreaterVisibilityToSymbol(Compilation compilation, ITypeSymbol type, ISymbol affectedSymbol)
+            {
+                var container = affectedSymbol.ContainingType;
+                while (container != null)
+                {
+                    if (!compilation.IsSymbolAccessibleWithin(affectedSymbol, container))
+                    {
+                        // the affected symbol is no longer visible, so we've passed the gauntlet
+                        return true;
+                    }
+
+                    if (!compilation.IsSymbolAccessibleWithin(type, container))
+                    {
+                        // if the type can't be reached here, we can't proceed
+                        return false;
+                    }
+
+                    container = container.ContainingType;
+                }
+
+                if (!compilation.IsSymbolAccessibleWithin(affectedSymbol, affectedSymbol.ContainingAssembly))
+                {
+                    // if the affected symbol is not visible at the assembly level, we're done
+                    return true;
+                }
+
+                // final check
+                return compilation.IsSymbolAccessibleWithin(type, affectedSymbol.ContainingAssembly);
             }
 
             bool CanUpgrade(IMethodSymbol methodSym) => !coll.MethodsAssignedToDelegate.ContainsKey(methodSym);

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs
@@ -2,6 +2,7 @@
 
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Testing;
+using Test.Utilities;
 using Xunit;
 using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
     Microsoft.NetCore.Analyzers.Performance.UseConcreteTypeAnalyzer,
@@ -12,6 +13,87 @@ namespace Microsoft.NetCore.Analyzers.Performance.UnitTests
     public static partial class UseConcreteTypeTests
     {
         [Fact]
+        [WorkItem(6565, "https://github.com/dotnet/roslyn-analyzers/issues/6565")]
+        public static async Task DiscoverArrayUpgrades()
+        {
+            await TestCSAsync(@"
+                using System;
+                using System.Collections.Generic;
+
+                public class X
+                {
+                    private static IList<string> {|#0:GetListPrivate|}()
+                    {
+                        return Array.Empty<string>();
+                    }
+                }
+            ", VerifyCS.Diagnostic(UseConcreteTypeAnalyzer.UseConcreteTypeForMethodReturn)
+                .WithLocation(0)
+                .WithArguments("GetListPrivate", "System.Collections.Generic.IList<string>", "string[]"));
+        }
+
+        [Fact]
+        [WorkItem(6687, "https://github.com/dotnet/roslyn-analyzers/issues/6687")]
+        public static async Task ShouldNotTrigger_ConflictingOverloads()
+        {
+            await TestCSAsync(@"
+                abstract class Base
+                {
+                    public virtual string M(object o) => ""M(object)"";
+                }
+
+                sealed class Derived : Base
+                {
+                    public void M(string s) {}
+                }
+
+                internal class C
+                {
+
+                    private readonly Base _a = new Derived();
+
+                    public void Trigger()
+                    {
+                        var s = _a.M("""");
+                    }
+                }
+            ", $"dotnet_code_quality.CA1859.api_surface = private,internal");
+        }
+
+        [Fact]
+        [WorkItem(6704, "https://github.com/dotnet/roslyn-analyzers/issues/6704")]
+        public static async Task ShouldNotTrigger_ExplicitInterfaceImplementation()
+        {
+            await TestCSAsync(@"
+                class Class1
+                {
+                    public void FalseWarning()
+                    {
+                        I obj = new Derived();
+                        obj.M();
+                    }
+                }
+
+                interface I
+                {
+                    void M();
+                }
+
+                class Base : I
+                {
+                    void I.M()
+                    {
+                    }
+                }
+
+                class Derived : Base
+                {
+                }
+            ", $"dotnet_code_quality.CA1859.api_surface = private,internal");
+        }
+
+        [Fact]
+        [WorkItem(6659, "https://github.com/dotnet/roslyn-analyzers/issues/6659")]
         public static async Task ShouldNotTrigger_Visibility()
         {
             await TestCSAsync(@"
@@ -32,6 +114,32 @@ namespace Microsoft.NetCore.Analyzers.Performance.UnitTests
                         Obj.ToString();
                         Test(new Nested());
                     }
+                }
+            ", $"dotnet_code_quality.CA1859.api_surface = private,internal");
+        }
+
+        [Fact]
+        public static async Task ShouldNotTrigger_VisibilityNestedTypes()
+        {
+            await TestCSAsync(@"
+                using System;
+
+                public class Program
+                {
+	                private static class NestedClass1
+	                {
+		                public static IDisposable Method()
+		                {
+			                return new NestedClass2();
+		                }
+
+		                private sealed class NestedClass2 : IDisposable
+		                {
+			                public void Dispose()
+			                {
+			                }
+		                }
+	                }
                 }
             ", $"dotnet_code_quality.CA1859.api_surface = private,internal");
         }

